### ISSUE TITLE
Option to output the context to a file

### ISSUE THREAD

**Issue by djbe** [SwiftGen/SwiftGen#310]
We might want to add a way for users to output the generated context from parsing file(s) to a plist file (like we use for testing).

This could be handy for template writers so that they have a reference of what's available, but also for us should we need a context when debugging an issue reported by a user.

**Comment by AliSoftware** [comment]
Yup 👍 

Just some thoughts:

* I’d prefer a `--print-context` option for example, i.e. sure a dedicated option but maybe not named like `--debug`, as I’m guessing it might also be used for usages other than just debugging — like integrating with 3rd party tools which would want to use that context for their own input

* I’m guessing if `--print-context` is used, then it would be mutually exclusive with the `-t/p` options (it’d be ether `-t` or `-p` or `-c`) But the `-o` option then would still be valid and used to redirect where the context is written, right?

**Comment by AliSoftware** [comment]
In case we chose the additive solution — i.e. that `-t/-p/-o`  flags can be used in addition to `-c` (aka `--print-context`) and that `-c` takes the path to the file to write the context into — and not the exclusive solution, here's a matrix of all possible choices.

This solution would allow people to both generate the output and print the context, or only print the context, or only generate the output. It would just be more complex to reason about all the possible combinations but nothing unmanageable so probably worth it

| `-t <nameT>` | `-p <pathP>` | `-o <pathO>` | `-c <pathC>` | Result |
|---------------|---------------|---------------|---------------|--------|
|    |    |    |    | ⛔️ Error: nothing to generate. Provide either a template to specify the code to generate or the `-c` option to generate the context file |
|    |    |    | -c | ✅ OK: Generates the context file only (in `pathC`). |
|    |    | -o |    | ⛔️ Error: nothing to generate. Provide either a template to specify the code to generate or the `-c` option to generate the context file |
|    |    | -o | -c | (Option A): ⛔️ Error: `-o` is useless without `-t` or `-p`. (Option B) ✅ OK: Generates the context file in `pathC`, disregard `-o <pathO>` (Option C) ⚠️ Both (warning, but generates the context file anyway) |
|    | -p |    |    | ✅ Generates the output file in stdout |
|    | -p |    | -c | ✅ Generates the output file in stdout and the Context file in `pathC` |
|    | -p | -o |    | ✅ Generates the output file in `pathO` |
|    | -p | -o | -c | ✅ Generates the output file in `pathO` and the Context file in `pathC` |
| -t |    |    |    | ✅ Generates the output file in stdout |
| -t |    |    | -c | ✅ Generates the output file in stdout and the Context file in `pathC` |
| -t |    | -o |    | ✅ Generates the output file in `pathO` |
| -t |    | -o | -c | ✅ Generates the output file in `pathO` and the Context file in `pathC` |
| -t | -p | * | * | ⛔️ Error: `-t` and `-p` are mutually exclusive, only provide one or the other |

---

The only big problem with that solution:
* Is there really a use case where people will want BOTH to generate the output file and the context file?
* In that solution, there's no real way to generate the context file in stdout to be piped through to a 3rd party tool, like `swiftgen <subcommand> -c <path_to_scan> | thirdpartytool`. Sure we could use `-c /dev/stdout` but that would be cheating and strange and would then allow to print both the generated file and the context file in stdout in such situations

---

Maybe a simpler solution would be to be exclusive, and allow only either `-t` or `-p` or `-c` but only one of them at a time, and use `-o` to specify the output path for all 3? Would allow us to:
* Be more simple about the flag combinations (easier documentation / usage docs)
* Allow users to dump the context on stdout to be piped thru any 3rd party tool without the need for an intermediate file
* In the future we might even imagine that if the `-c`/`--context` flag is set without `-t`/`-p` it means "dump the context" but if it's used alongside `-t`/`p` it would mean "interpret the input path provided as a context file already parsed" thus allowing `swiftgen strings -t swift3 -c colors-context.plist -o colors.swift`. This would allow user to split the process in 2 steps (generating context first, inspecting it, then generating the output _from_ an existing context next, without restarting the parsing from scratch). Might also help in tweaking templates with example contexts in an easier way for template authors.

| `-t <nameT>` | `-p <pathP>` | `-o <pathO>` | `-c` | Result |
|---------------|---------------|---------------|---------------|--------|
|    |    |    |    | ⛔️ Error: nothing to generate. Provide either a template to specify the code to generate or the `-c` option to generate the context file |
|    |    |    | -c | ✅ OK: Generates the context in stdout. |
|    |    | -o |    | ⛔️ Error: nothing to generate. Provide either a template to specify the code to generate or the `-c` option to generate the context file |
|    |    | -o | -c | ✅ OK: Generates the context in `pathO`. |
|    | -p |    |    | ✅ Generates the output file in stdout |
|    | -p |    | -c | (Option A) ⛔️ Error: `-t`/`-p`/`c` are mutually exclusive. (Option B) ✅ Interpret the input file as a context file rather than a resource to parse, and generate the output file in stdout from that context. |
|    | -p | -o |    | ✅ Generates the output file in `pathO` |
|    | -p | -o | -c | (Option A) ⛔️ Error: `-t`/`-p`/`c` are mutually exclusive. (Option B) ✅ Interpret the input file as a context file rather than a resource to parse, and generate the output file in `pathO` from that context. |
| -t |    |    |    | ✅ Generates the output file in stdout |
| -t |    |    | -c | (Option A) ⛔️ Error: `-t`/`-p`/`c` are mutually exclusive. (Option B) ✅ Generates the output file in stdout from the context file provided as input |
| -t |    | -o |    | ✅ Generates the output file in `pathO` |
| -t |    | -o | -c | (Option A) ⛔️ Error: `-t`/`-p`/`c` are mutually exclusive. (Option B) ✅ Generates the output file in `pathO` from the context file provided as input |
| -t | -p | * | * | ⛔️ Error: `-t` and `-p` are mutually exclusive, only provide one or the other |

This way I think the rules are easier to understand for the user, while still allowing them to pipe the context to a third-party tool. And if we choose Option B we could even allow users to use a context file as input if they want to. Could allow powerful stuff like `swiftgen strings -c Localizable.strings | filter-and-reshape-strings-context-file | swiftgen strings -c -t swift3 -o strings-generated.swift`

**Comment by AliSoftware** [comment]
Re-reading that I think I more and more like the 2nd approach better.

We could imagine implement a first version (option A) to have it quickly available (e.g. in 5.1) — only allowing `-c -o path` to output the context in path (and throwing an error if `-t/-p` is used with `-c`) — and a second version later —allowing `-c -t name -o path input` to implement Option B when we have more time

**Comment by djbe** [comment]
Yeah, 2nd approach, option A for now. Option B should be doable once A is there, maybe even in 5.1 or 5.2.
